File: .tmp_py4013/batch_04_sources/2476_closest_nodes_queries_in_a_binary_search_tree.js
// LeetCode 2476 - Closest Nodes Queries in a Binary Search Tree
// https://leetcode.com/problems/closest-nodes-queries-in-a-binary-search-tree/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number[]} queries
 * @return {number[][]}
 */
var closestNodes = function(root, queries) {
    const vals = [];
    const inorder = (node) => {
        if (!node) return;
        inorder(node.left);
        vals.push(node.val);
        inorder(node.right);
    };
    inorder(root);
    const lowerBound = (q) => {
        let lo = 0, hi = vals.length;
        while (lo < hi) {
            const mid = (lo + hi) >> 1;
            if (vals[mid] < q) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    };
    const ans = [];
    for (const q of queries) {
        const j = lowerBound(q);
        const mx = j < vals.length ? vals[j] : -1;
        let mn = -1;
        if (j < vals.length && vals[j] === q) mn = q;
        else if (j > 0) mn = vals[j - 1];
        ans.push([mn, mx]);
    }
    return ans;
};

File: .tmp_rs4013/rust_check_batch08/2476_closest_nodes_queries_in_a_binary_search_tree.rs
struct Solution;
// LeetCode 2476 - Closest Nodes Queries in a Binary Search Tree
// https://leetcode.com/problems/closest-nodes-queries-in-a-binary-search-tree/

use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn closest_nodes(
        root: Option<Rc<RefCell<TreeNode>>>,
        queries: Vec<i32>,
    ) -> Vec<Vec<i32>> {
        let mut vals = Vec::new();
        fn inorder(node: &Option<Rc<RefCell<TreeNode>>>, vals: &mut Vec<i32>) {
            if let Some(n) = node {
                let n = n.borrow();
                inorder(&n.left, vals);
                vals.push(n.val);
                inorder(&n.right, vals);
            }
        }
        inorder(&root, &mut vals);
        let mut ans = vec![vec![0; 2]; queries.len()];
        for (i, &q) in queries.iter().enumerate() {
            let j = vals.partition_point(|&x| x < q);
            let mx = if j < vals.len() { vals[j] } else { -1 };
            let mn = if j < vals.len() && vals[j] == q {
                q
            } else if j > 0 {
                vals[j - 1]
            } else {
                -1
            };
            ans[i] = vec![mn, mx];
        }
        ans
    }
}

fn main() {}
